File: extension/graphkit/services/graph-service.ts
// GraphService: high-level graph operations.
//
// Phase 2 service layer. Reads/writes flow through the Phase 1 storage layer
// (storage/graph-storage.ts). Reads degrade gracefully (never throw); writes
// surface failures so callers can react. Full graph generation and FTS-backed
// related-code search are deferred to later phases — for now generateGraph
// returns any existing graph (or empty) and findRelatedCode returns an empty
// list, which preserves current behavior without crashing.
//
// NOTE on brief drift: the task brief referenced `node.label` and a `mode`
// field on GraphData, and imported `rollupCodeRefs`/`renderRepoMemory` from
// graph.mjs. None of those match the shipped Phase 1 surface:
//   - GraphNode uses `title` (not `label`); querying on `label` would never
//     match real nodes, so this implementation searches `title`.
//   - GraphData is `{ nodes, edges }` with no `mode` field, so none is emitted
//     on returned objects (the `mode` parameter is still accepted on
//     generateGraph for API stability / future use).
//   - graph.mjs does not export `renderRepoMemory`, and `rollupCodeRefs` is not
//     used by the stubbed findRelatedCode; importing a non-existent named
//     binding would crash the module at load time, so no graph.mjs import is
//     taken here. Delegation will be wired in a later phase alongside FTS.

import {
  readGraphFile,
  writeDocFingerprint
} from '../storage/graph-storage.ts';
import type { GraphData, GraphNode, GraphMode, CodeRef } from '../types/graph.ts';

/**
 * GraphService provides high-level graph operations.
 *
 * Initially delegates to the storage layer for parity; richer generation and
 * FTS search arrive in later phases.
 */
export class GraphService {
  /**
   * Generate graph for a repo in the specified mode.
   *
   * For now this returns any existing graph on disk, or an empty graph for a
   * fresh repo. Full generation lands in a later phase.
   */
  async generateGraph(repoRoot: URL, mode: GraphMode = 'code'): Promise<GraphData> {
    try {
      const existing = await readGraphFile(repoRoot);
      if (existing.nodes.length > 0 || existing.edges.length > 0) {
        return existing;
      }
      return { nodes: [], edges: [] };
    } catch (err) {
      // Graceful degradation — never crash the agent.
      console.error(`Graph generation failed (mode=${mode}):`, err);
      return { nodes: [], edges: [] };
    }
  }

  /**
   * Query graph for nodes whose title contains the query string
   * (case-insensitive). Returns up to `limit` matches.
   */
  async queryGraph(repoRoot: URL, query: string, limit = 10): Promise<GraphNode[]> {
    try {
      const graph = await readGraphFile(repoRoot);
      const q = typeof query === 'string' ? query.toLowerCase() : '';
      if (!q) return [];

      const results = graph.nodes.filter((node) => {
        const title = typeof node.title === 'string' ? node.title.toLowerCase() : '';
        return title.includes(q);
      });

      return results.slice(0, limit);
    } catch (err) {
      console.error('Graph query failed:', err);
      return [];
    }
  }

  /**
   * Find code related to a query. Stubbed to an empty list for now; full FTS
   * search (delegating to graph.mjs rollupCodeRefs over the KB) lands in a
   * later phase. Returning empty is the safe non-crashing default.
   */
  async findRelatedCode(repoRoot: URL, query: string, limit = 10): Promise<CodeRef[]> {
    try {
      // TODO: Implement full FTS search in later phases (delegating to
      // graph.mjs rollupCodeRefs over the KB). Returning empty is the safe
      // non-crashing default; `limit` is honored now so the eventual
      // implementation inherits the cap.
      const results: CodeRef[] = [];
      return results.slice(0, limit);
    } catch (err) {
      console.error('Related code search failed:', err);
      return [];
    }
  }

  /**
   * Regenerate graph. Currently just refreshes the doc fingerprint so callers
   * can detect subsequent drift; full regeneration lands in a later phase.
   *
   * Unlike reads, a failed fingerprint write is re-thrown after logging so
   * callers can surface the failure rather than silently corrupting state.
   */
  async regenerateGraph(repoRoot: URL): Promise<void> {
    try {
      await writeDocFingerprint(repoRoot, Date.now().toString());
    } catch (err) {
      console.error('Graph regeneration failed:', err);
      throw err;
    }
  }
}

// Singleton instance.
export const graphService = new GraphService();

File: mac/Sources/LlmIdeMac/Services/GraphService.swift
// High-level graph service: typed read/query over the Phase 1 GraphStorage
// layer. This is the service-tier surface other Mac-app code (and later
// service tasks) consume; it delegates all file I/O to GraphStorage and adds
// graceful degradation (never crashes on read — returns empty data / logs).
//
// Mirrors the TS service layer (Task 5 of the storage/service split). The
// `findRelatedCode` FTS search and true (re)generation are forward-looking
// placeholders landing in a later phase — `findRelatedCode` returns `[]` and
// `generateGraph` reads the existing on-disk graph (or returns an empty one)
// rather than running the scanners itself; `regenerateGraph` simply bumps the
// doc fingerprint so the auto-updater picks up a rebuild.
//
// Deviations from the original task template (forced by the real Phase 1
// types so the code compiles with no breaking changes):
//   - The "GraphData"/"GraphNode" the template names are GraphKit's `CGData`
//     and `CGNode`. They are re-used here (no aliases) to match GraphStorage
//     and the rest of the module.
//   - `CGNode` exposes `title` (not `label`), so `queryGraph` searches
//     `node.title`. `CGData` carries no `mode` field, so `mode` is not stored
//     on the returned graph (it only selects code vs. doc for future use).
//   - `GraphMode` does not exist in the module today, so a small
//     `Sendable`/`Codable` enum is declared here (`.code` / `.doc`) matching
//     the dual code/doc graph model already in `KnowledgeGraphService`.
//   - `CodeRef` already exists (`Models/Plan.swift`) with a richer shape
//     (`ref?`/`title`/`bodyExcerpt?`/`rank?`) than the template's
//     `ref/snippet/score`. It is reused as-is — the template explicitly notes
//     "CodeRef struct (if not already defined)" — so `findRelatedCode`
//     returns the existing type, avoiding a duplicate-declaration conflict.
//   - `GraphStorage` methods label their first parameter `repoRoot:`, so all
//     delegation calls pass it explicitly (the template's positional calls
//     would not compile).

import Foundation
import GraphKit

/// Which graph a `GraphService` operation targets. Matches the dual code/doc
/// graph model in `KnowledgeGraphService`. `Sendable` & `Codable` so it can
/// cross actor boundaries and round-trip alongside graph requests.
enum GraphMode: String, Sendable, Codable {
    case code
    case doc
}

/// Graph service protocol for high-level graph operations.
///
/// Conforming types must be `Sendable` (the bundled impl is an actor) so a
/// single shared instance is safe to reuse across the app.
protocol GraphService: Sendable {
    /// Read the existing on-disk graph for a repo, returning an empty graph
    /// when none exists (or when read fails). `mode` selects code vs. doc for
    /// future phases; today the on-disk `graph.json` is returned regardless.
    func generateGraph(repoRoot: URL, mode: GraphMode) async throws -> CGData

    /// Simple case-insensitive text search over node titles. Returns up to
    /// `limit` matching nodes; `[]` on read failure.
    func queryGraph(repoRoot: URL, query: String, limit: Int) async throws -> [CGNode]

    /// Find code references related to `query`. Placeholder for full FTS in a
    /// later phase — returns `[]` today. Reuses the existing `CodeRef` type.
    func findRelatedCode(repoRoot: URL, query: String, limit: Int) async throws -> [CodeRef]

    /// Mark the repo's graph for regeneration by writing a fresh doc
    /// fingerprint (the auto-updater rebuilds on the next change cycle).
    func regenerateGraph(repoRoot: URL) async throws
}

/// Graph service implementation. An actor for thread safety; all file I/O is
/// delegated to the injected `GraphStorage` (Phase 1 storage layer).
final actor GraphServiceImpl: GraphService {
    private let storage: GraphStorage

    init(storage: GraphStorage = GraphStorage()) {
        self.storage = storage
    }

    func generateGraph(repoRoot: URL, mode: GraphMode = .code) async throws -> CGData {
        do {
            // Try reading existing graph first; if it has any content, return
            // it as-is. Otherwise return an empty graph for the requested mode.
            let existing = try await storage.readGraphFile(repoRoot: repoRoot)
            if !existing.nodes.isEmpty || !existing.edges.isEmpty {
                return existing
            }
            return CGData.empty
        } catch {
            // Graceful degradation — never crash; log and return empty.
            print("Graph generation failed: \(error)")
            return CGData.empty
        }
    }

    func queryGraph(repoRoot: URL, query: String, limit: Int = 10) async throws -> [CGNode] {
        do {
            let graph = try await storage.readGraphFile(repoRoot: repoRoot)

            // Simple case-insensitive text search over node titles.
            let results = graph.nodes.filter { node in
                node.title.localizedCaseInsensitiveContains(query)
            }

            return Array(results.prefix(limit))
        } catch {
            // Graceful degradation — log and return empty rather than throw.
            print("Graph query failed: \(error)")
            return []
        }
    }

    func findRelatedCode(repoRoot: URL, query: String, limit: Int = 10) async throws -> [CodeRef] {
        // TODO: Implement full FTS search in later phases.
        return []
    }

    func regenerateGraph(repoRoot: URL) async throws {
        // Write a new fingerprint to mark the graph for regeneration; the
        // auto-updater compares this against the live doc set and rebuilds.
        try await storage.writeDocFingerprint(
            repoRoot: repoRoot, fingerprint: String(Date().timeIntervalSince1970))
    }
}
